File: OptimalPos.py
import dsp
import cvxpy as cp
import numpy as np
from Models import BG
from Distortions import minmaxvar as mmv
import matplotlib.pyplot as plt

params = [0.04,10/52,1,5/52,0.04,10/52,1,5/52]

k = 8
W = 1
M = [-1,1]
model = BG(k,M,params)
p = model.p
q = model.q
x = model.y

# fig = plt.figure()
# axes = fig.add_axes([0.1, 0.1, 1, 1])
# axes.set_xlim(M[0], M[1])
# axes.set_ylim(0, max([max(p),max(q)]))
# axes.plot(x, p)
# axes.plot(x, q)
# plt.show()

lam = 5
a = np.linspace(0.1,5,100)
dist = mmv(lam)
Phi = dist.Phi(a)
# fig = plt.figure()
# axes = fig.add_axes([0.1, 0.1, 1, 1])
# axes.set_xlim(a[0]-0.5, a[-1]+0.5)
# axes.set_ylim(0, max(Phi))
# axes.plot(a, Phi)
# # axes.plot(a,dist.Psi(a))
# plt.show()

theta = 0.75
alpha = 1.25
beta = 0.25
P = np.diag(p)

y = cp.Variable(2**k)
z = cp.Variable(2**k)

f = dsp.inner(z, P @ (y - cp.multiply(W,cp.exp(x))))
rho = p @ (cp.multiply(theta, cp.power(z,alpha))+cp.multiply(1-theta,cp.power(z,-beta)))
obj = dsp.MinimizeMaximize(rho-f)
constraints = [q @ y == W, p @ z == 1, z >= 0]
for i in range(89): #range(len(a)):
    constraints.append(p @ cp.maximum(z-a[i],0) <= Phi[i])

'''
Note: too many constraints generate some problems in CVXPY, see https://github.com/cvxpy/cvxpy/issues/826
Because of this I have here limited the number of constraints. It can be seen that the solution z does satisfy the constraint.

A potential solution is suggested on github:
The problem is that when calculating the shape for scipy.sparse.csc_matrix, it always gets a negative value,
since the shape is numpy.int32 and if the shape is too large, we get overflow problem. The idea to solve it is very simple,
just convert it into numpy.int64.
So do the following steps:
1. find canonInterface.py in cvxpy\cvxcore\python and find get_problem_matrix function. you will see around 368line, use this
instead A = scipy.sparse.csc_matrix((V, (I, J)), shape=(np.int64(constr_length)*np.int64(var_length+1), param_size_plus_one))
2. find conic_solver.py in \cvxpy\reductions\solvers\conic_solvers and find format_constraints function. 
You will see around 225 line and use this instead:
restructured_A = 
    restructured_A.reshape(np.int64(restruct_mat.shape[0]) * np.int64(problem.x.size + 1),problem.A.shape[1], order='F')
'''

obj = dsp.MinimizeMaximize(f)
prob = dsp.SaddlePointProblem(obj, constraints)
prob.solve()  # solves the problem

print(prob.value)

fig = plt.figure()
axes = fig.add_axes([0.1, 0.1, 0.75, 0.75])
axes.set_xlim(np.log(W)+M[0], np.log(W) + M[-1])
axes.set_ylim(min(y.value), max(y.value))
axes.plot(x,y.value)
# axes.plot(a,dist.Psi(a))
plt.show()

# Constraint satisfied
# print(q @ y.value)
zz = z.value
N = len(a)
const = []
for i in range(N): #range(len(a)):
    const.append(p @ np.maximum(zz-a[i],0))

const = np.array(const)
fig = plt.figure()
axes = fig.add_axes([0.1, 0.1, 0.75, 0.75])
axes.set_xlim(a[0], a[-1])
axes.set_ylim(min(const), max(Phi))
axes.plot(a,const)
axes.plot(a,Phi)
# axes.plot(a,dist.Psi(a))
plt.show()

File: OptimalPos.m
clear
clc

%% Parameters

% BG parameters
params = [0.04,10/52,1,5/52,0.02,10/52,1.5,5/52];
bp = params(1);
cp = params(2);
bn = params(3);
cn = params(4);

% Discretization
k = 8;
M = [-0.5,0.5];
y = linspace(M(1),M(2),2^k);
ip = (y > 0);
im = (y < 0);
yp = y(ip);
ym = -y(im);
delta = (M(2)-M(1))/(2^k-1);

b = (1/bp+1/bn);
gammap = gamma(cp);
gamman = gamma(cn);
lam = 0.5*(cp+cn);
mu = 0.5*(cp+cn-1);
x = y*b;
xp = x(ip);
xm = -x(im);

W = zeros(size(x));
p = zeros(size(x));
W(ip) = exp(-0.5*xp) .* (xp.^(mu+0.5)) .* kummerU(mu-lam+0.5, 1+2*mu, xp);
W(im) = exp(-0.5*xm) .* (xm.^(mu+0.5)) .* kummerU(mu-lam+0.5, 1+2*mu, xm);
p(ip) = ( (bp)^(-cp) ) * ( (bn)^(-cn) ) * ( (yp).^(0.5*(cp+cn)-1) ) .* exp(-0.5*yp) .* W(ip) / gammap;
p(im) = ( (bp)^(-cp) ) * ( (bn)^(-cn) ) * ( (ym).^(0.5*(cp+cn)-1) ) .* exp(-0.5*ym) .* W(im) / gamman;
p = p * delta / ( b^(0.5*(cp+cn)) );
p = p/sum(p);

% BG Risk neutral parameters
bptil = params(5);
cptil = params(6);
bntil = params(7);
cntil = params(8);

% Discretization
k = 8;
M = [-0.5,0.5];

b = (1/bptil+1/bntil);
gammap = gamma(cptil);
gamman = gamma(cntil);
lam = 0.5*(cptil+cntil);
mu = 0.5*(cptil+cntil-1);
x = y*b;
xp = x(ip);
xm = -x(im);

W = zeros(size(x));
p = zeros(size(x));
W(ip) = exp(-0.5*xp) .* (xp.^(mu+0.5)) .* kummerU(mu-lam+0.5, 1+2*mu, xp);
W(im) = exp(-0.5*xm) .* (xm.^(mu+0.5)) .* kummerU(mu-lam+0.5, 1+2*mu, xm);
q(ip) = ( (bptil)^(-cptil) ) * ( (bntil)^(-cntil) ) * ( (yp).^(0.5*(cptil+cntil)-1) ) .* exp(-0.5*yp) .* W(ip) / gammap;
q(im) = ( (bptil)^(-cptil) ) * ( (bntil)^(-cntil) ) * ( (ym).^(0.5*(cptil+cntil)-1) ) .* exp(-0.5*ym) .* W(im) / gamman;
q = q * delta / ( b^(0.5*(cptil+cntil)) );
q = q/sum(p);

% Constraints
N = 50;
a = linspace(0.5,2,N);
lam = 0.25;
Phi = MMV_Phi(a,lam);

% Rebate
theta = 0.25;
P = diag(p);

%% Optimization

% load python libraries

pyenv('Version',...
    'C:\Users\yoshi\OneDrive\Desktop\Research\OptimalDerivativePos\Maximin\maximin2\python.exe',...
    'ExecutionMode','OutOfProcess');
py.importlib.import_module('dsp');
py.importlib.import_module('cvxpy');
py.importlib.import_module('numpy');

l_mat = rand(2,2);
l_py = py.numpy.random.random([int64(2), int64(2)]);

npA = py.numpy.array(l_mat(:).');

p = py.numpy.array(p.');
q = py.numpy.array(q.');
x = py.numpy.array(y.');

P = py.numpy.diag(p);

pyrun("f = dsp.inner(z, P @ (y - q @ y))", " f_mat");
% rho = p @ (cp.multiply(theta, cp.power(z,alpha))+cp.multiply(1-theta,cp.power(z,-beta)))
% obj = dsp.MinimizeMaximize(rho+f)
% constraints = [p @ z == 1, z >= 0]
% for i in range(N): #range(len(a)):
%     constraints.append(p @ cp.maximum(z-a[i],0) <= Phi[i])

%pyrun("l = py.numpy.array(l)")

%% Routines

function psi = MMV_Psi(x,lam)
    psi = 1 - ( 1-x.^(1/(lam+1)) ).^(1+lam);
end

function phi = MMV_Phi(a,lam)
    phi = ( 1 ./ a.^(1/lam) ).^(lam+1);
end